### PR TITLE
feat: query multi-version cells

### LINKED ISSUE #11: Query multiple versions of Spore/Cluster cells
Currently, spore-graphql only queries the latest version of the Spore/Cluster cells, filtered with the getSporeScript API. 
However, in reality, there could be multiple versions (v1/v2) of the scripts that need to be queried in the list.

### PR DESCRIPTION
**Pull request by ShookLyngs**
## Changes
- Added support of multi-version spores/clusters querying
- Bumped lib/example dependencies (spore-sdk, lumos)

## Checklist
- [ ] Need more tests: https://github.com/sporeprotocol/spore-graphql/issues/11#issuecomment-1920437756

## Related issues
- resolves #11

### CODE REVIEW

**Comment by ashuralyk** [comment]
@ShookLyngs I've pushed some commits to solve problems you mentioned, please take a look

**Comment by ShookLyngs** [comment]
- The query for spores/clusters version-by-version seems fine to me. The only downside is that users of older versions may not be satisfied with the querying logic, as new version spores can easily overshadow them. However, I'm not sure how it will perform in the real world. It would be a good idea to test it locally with the spore-graphql example.
- The added filter inputs, "codeHashes," are still not handled in the query logic. Do you think they should be removed for now? Ref to: https://github.com/sporeprotocol/spore-graphql/pull/12/files#diff-160c34fe61acb6edcfc53a821f18efc0b0aa49a3a92582992f17338613d7204cR16-R27

**Comment by ashuralyk** [comment]
> * The added filter inputs, "codeHashes," are still not handled in the query logic. Do you think they should be removed for now? Ref to: https://github.com/sporeprotocol/spore-graphql/pull/12/files#diff-160c34fe61acb6edcfc53a821f18efc0b0aa49a3a92582992f17338613d7204cR16-R27

I know the not used `codeHashes` field, it's just added in graphql schema but not added in codebase, right?

> The only downside is that users of older versions may not be satisfied with the querying logic, as new version spores can easily overshadow them

we searched all of Spores with different version, and export interfaces leverage with `code-hash` which can distinguish Spore version, so I just think this won't be a problem?

File: lib/src/resolvers/cluster.ts [hunk 1]
```diff
@@ -85,8 +93,13 @@ export async function getMintableClusters(
 /**
  * Get the count of clusters
  */
-export async function getClusterCount(_: unknown, __: unknown, { dataSources }: ContextValue): Promise<number> {
-  const key: ClusterLoadKeys = ['0x', 'desc', Number.MAX_SAFE_INTEGER, undefined];
+export async function getClusterCount(
+  _: unknown,
+  __: unknown,
+  { dataSources }: ContextValue,
+  { codeHash }: { codeHash?: string },
```
Comment (ShookLyngs) [inline lib/src/resolvers/cluster.ts]: Comparing with other functions, the position of the filter inputs argument seems wrong.

File: lib/src/resolvers/cluster.ts [hunk 2]
```diff
@@ -43,10 +43,18 @@ export async function getTopClusters(
   { dataSources }: ContextValue,
 ): Promise<Cluster[]> {
   const { first, after, filter } = getQueryParams(params);
-  const { mintableBy } = filter ?? {};
+  const { mintableBy, codeHash } = filter ?? {};
```
Comment (ShookLyngs) [inline lib/src/resolvers/cluster.ts:53]: "codeHash" was not defined in the "Query.topClusters" in the schema. 
You should add it to the "TopClusterFilterInput" type.

File: lib/src/resolvers/cluster.ts [hunk 3]
```diff
@@ -68,10 +76,10 @@ export async function getTopClusters(
  */
 export async function getMintableClusters(
   _: unknown,
-  { address }: { address: string },
+  { address, codeHash }: { address: string; codeHash?: string },
```
Comment (ShookLyngs) [inline lib/src/resolvers/cluster.ts]: "codeHash" was not defined in the "Query.mintableClusters" type in the schema.
You should define a input filter type for the function and add it to the "Query.mintableClusters" type.

**Review by ashuralyk — APPROVED** [review]

**Pull request merged**